# Lab 01: Data wrangling and EDA

**Name: Maxwell Cooper**

## Question  1.1

In [57]:
import numpy as np
import pandas as pd
import plotly.express as px

In [2]:
gifts=pd.read_csv('ForeignGifts_edu.csv')

In [5]:
gifts.head()

,ID,OPEID,Institution Name,City,State,Foreign Gift Received Date,Foreign Gift Amount,Gift Type,Country of Giftor,Giftor Name
0,1,102000,Jacksonville State University,Jacksonville,AL,43738,250000,Monetary Gift,CHINA,NaN
1,2,104700,Troy University,Troy,AL,43592,463657,Contract,CHINA,Confucius Institute Headquarters
2,3,105100,University of Alabama,Tuscaloosa,AL,43466,3649107,Contract,ENGLAND,Springer Nature Customer Service Ce
3,4,105100,University of Alabama,Tuscaloosa,AL,43472,1000,Contract,SAUDI ARABIA,Saudi Arabia Education Mission
4,5,105100,University of Alabama,Tuscaloosa,AL,43479,49476,Contract,SAUDI ARABIA,Saudi Arabia Education Mission


## Question 1.2

In [18]:
print(gifts['Gift Type'].value_counts())
print(gifts['Gift Type'].value_counts(normalize=True))

Gift Type
Contract         17274
Monetary Gift    10936
Real Estate         11
Name: count, dtype: int64
Gift Type
Contract         0.612097
Monetary Gift    0.387513
Real Estate      0.000390
Name: proportion, dtype: float64


>There are three types of gifts given: Contracts, Monetary gifts, and real estate. Most of the gifts were contracts or longer term gifts with 17272 contributions totaling to about 61% of all gifts. The second most common were one time monetary gifts with 10936 gifts or about 39%. The least common gift type was real estate in the different country at only 11 out of the 28221 total gifts or .03% of all gifts.

## Q 1.3.1

In [34]:
total_money=gifts.groupby('Country of Giftor')['Foreign Gift Amount'].sum()
pd.DataFrame(total_money).sort_values(by='Foreign Gift Amount', ascending=False)


,Foreign Gift Amount
Country of Giftor,
QATAR,2706240869
ENGLAND,1464906771
CHINA,1237952112
SAUDI ARABIA,1065205930
BERMUDA,899593972
...,...
IRAN,80
ST. LUCIA,50
CROATIA,26


>The country that gave the most total money is Qatar.

## Q 1.3.2

In [ ]:
gifts['Country of Giftor'].value_counts()

Country of Giftor
ENGLAND                  3655
CHINA                    2461
CANADA                   2344
JAPAN                    1896
SWITZERLAND              1676
                         ... 
ROMANIA                     1
TUNISIA                     1
ST. KITTS-NEVIS             1
MARSHAL ISLANDS (THE)       1
ANTIGUA                     1
Name: count, Length: 155, dtype: int64

>The country that initiates the most number of gifts is England with 3655 total gifts given.

## Q 1.3.3

In [41]:
average_money=gifts.groupby('Country of Giftor')['Foreign Gift Amount'].mean()
pd.DataFrame(average_money).sort_values(by='Foreign Gift Amount', ascending=False)


,Foreign Gift Amount
Country of Giftor,
BERMUDA,7.688837e+06
ECUADOR,3.985140e+06
QATAR,3.905109e+06
ANTIGUA,3.435545e+06
JERSEY,3.064589e+06
...,...
ST. LUCIA,5.000000e+01
IRAN,4.000000e+01
CROATIA,2.600000e+01


>The Country of Bermuda gave the largest gifts on average with the average donation size of $7,688,837

## Q 1.3.4

In [44]:
gifts.groupby('Institution Name').sum().sort_values(by='Foreign Gift Amount', ascending=False)['Foreign Gift Amount']

Institution Name
Carnegie Mellon University               1477922504
Cornell University                       1289937761
Harvard University                        954803610
Massachusetts Institute of Technology     859071692
Yale University                           613441311
                                            ...    
Jacksonville State University                250000
Metropolitan State University                235439
University of California, Merced             182514
University of Central Oklahoma               160288
University of Jamestown                         500
Name: Foreign Gift Amount, Length: 318, dtype: int64

>Carnegie Mellon University received the most amount of money in foreign gifts with a whopping 1.4 Billion dollars received in total!

## Q 1.3.5

In [49]:
gifts['Institution Name'].value_counts()

Institution Name
University of California, Los Angeles          3916
California Institute of Technology             2502
Columbia University in the City of New York    1127
Ohio State University (The)                    1014
Johns Hopkins University                        899
                                               ... 
University of Wisconsin - Platteville             1
Des Moines Area Community College                 1
Beth Medrash Govoha of America                    1
Columbia Southern University                      1
Presidio Graduate School                          1
Name: count, Length: 318, dtype: int64

>UCLA received the most number of donations with 3916 total gifts given

In [ ]:
x=gifts['Institution Name'].value_counts()
x=pd.DataFrame(x)
x

,count
Institution Name,
"University of California, Los Angeles",3916
California Institute of Technology,2502
Columbia University in the City of New York,1127
Ohio State University (The),1014
Johns Hopkins University,899
...,...
University of Wisconsin - Platteville,1
Des Moines Area Community College,1
Beth Medrash Govoha of America,1


In [59]:
px.bar(x[:20])

>Here is a plotly bar chart showing the top 20 institutions by number of gifts received.

## Q 1.3.6

In [66]:
university_totals=pd.DataFrame(gifts.groupby('Institution Name').sum().sort_values(by='Foreign Gift Amount', ascending=False)['Foreign Gift Amount'])

mean_gift=university_totals['Foreign Gift Amount'].mean()
median_gift=university_totals['Foreign Gift Amount'].median()

print(f'The mean gift size given to a university is ${round(mean_gift,2)} whereas the median gift size given to a country is ${round(median_gift,2)}')

The mean gift size given to a university is $52202878.97 whereas the median gift size given to a country is $6486791.5


>The mean gift size is much larger than the median gift size. The mean is over 52 million but the median is around 6.5 million. 
>This 9 times increase is most likely due to a large skew to the right for some universities receiving lots of big donations. Most universities receive small gifts.

In [67]:
px.histogram(university_totals['Foreign Gift Amount'])

>This plot confirms our suspicions of very large skews to the right.

## Q 1.3.7

In [72]:
gift_crosstab=pd.crosstab(gifts['Country of Giftor'],gifts['Institution Name'],
            values=gifts['Foreign Gift Amount'],
            aggfunc=sum)
gift_crosstab

Institution Name,Adelphi University,Albert Einstein College of Medicine,Alfred University,American University (The),Amherst College,Arizona State University,Auburn University Montgomery,Babson College,Ball State University,Barnard College,...,Wilkes University,William Marshall Rice University,Williams College,Winthrop University,Worcester Polytechnic Institute,Wright State University,Xavier University of Louisiana,Yale University,Yeshiva University,Young Americans College of the Performing Arts (The)
Country of Giftor,,,,,,,,,,,,,,,,,,,,,
AFGHANISTAN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
AMERICAN SAMOA,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
ANGOLA,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
ANTIGUA,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
ARGENTINA,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
VIRGIN ISLANDS (BRITISH),NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,2858666.0,NaN,NaN
WALES,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,NaN,720000.0,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
YEMEN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN


In [71]:
import plotly.graph_objects as go

df = pd.read_csv('ForeignGifts_edu.csv',low_memory=False)
df.head()

#giftor = 'Giftor Name'
giftor = 'Country of Giftor'
recipi = 'Institution Name'
flow = 'Foreign Gift Amount'
N = 25

flows = (
    df.groupby([giftor, 
                recipi])
      [flow]
      .sum()
      .nlargest(N)
      .reset_index()
)

labels = (
    flows[giftor].tolist()
    + flows[recipi].tolist()
)

labels = list(dict.fromkeys(labels))

fig = go.Figure(
    go.Sankey(
        node=dict(label=labels),
        link=dict(
            source=flows[giftor]
                        .map(labels.index),
            target=flows[recipi]
                        .map(labels.index),
            value=flows[flow]
        )
    )
)

fig.show()

>The alluvial plot can also show us a lot of what we need. While, the plot itself does not give a concrete answer, it does give us a clear top choices to analyze. We can see that some of the largest flows are from Qatar to Cornell, Texas A&M, and Carnegie Mellon and Bermuda has a large flow to Carnegie Mellon. Now, we can find out which is the largest.

In [99]:
gift_crosstab.loc['QATAR','Carnegie Mellon University']
#largest_gifts={'Qatar to CMU':gift_crosstab.loc['QATAR','Carnegie Mellon University'],
#              'Qatar to Texas A&M': gift_crosstab.loc['QATAR','Texas A&M University'],
#              'Qatar to Cornell':gift_crosstab.loc['QATAR','Cornell University'],
#              'Bermuda to CMU':gift_crosstab.loc['BERMUDA','Carnegie Mellon University']}
#largest_gifts=pd.DataFrame(largest_gifts)

pd.DataFrame([gift_crosstab.loc['QATAR','Carnegie Mellon University'],gift_crosstab.loc['QATAR','Texas A&M University'],
            gift_crosstab.loc['QATAR','Cornell University'],gift_crosstab.loc['BERMUDA','Carnegie Mellon University']],
            index=['Qatar to CMU','Qatar to Texas A&M','Qatar to Cornell','Bermuda to CMU']).rank(0, ascending=False)

,0
Qatar to CMU,4.0
Qatar to Texas A&M,3.0
Qatar to Cornell,1.0
Bermuda to CMU,2.0


>Now we get our concrete answer from the our prediction based on the graph. Qatar to Cornell is the largest flow of money followed by Bermuda to CMU, Qatar to Texas A&M, and then Qatar to CMU.

## Q 1.3.8

In [105]:
pd.DataFrame(gifts['Foreign Gift Amount'].groupby(gifts['Giftor Name']).sum()).sort_values(by='Foreign Gift Amount', ascending=False)

,Foreign Gift Amount
Giftor Name,
Qatar Foundation,1166503744
Qatar Foundation/Qatar National Res,796197000
Qatar Foundation for Education,373945215
Anonymous,338793629
Saudi Arabian Cultural Mission,275221475
...,...
"Noh, Jongyoun",1
"Zhang, Stephanie",1
"VISETSUTH, PEERANUT",1


>The top gift givers are the Qatar Foundation, the Qatar Foundation/Qatar National Res, the Qatar Foundation for Education, and the Saudi Arabian Cultural Mission. Funnily enough, the 4th highest gift giver belongs to 'Anonymous' although this is most likely a collection of many people across multiple countries.